### PR TITLE
Update handling of collection literals with no target type

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs [hunk 1]
```diff
@@ -60,6 +60,7 @@ internal bool NeedsToBeConverted()
                 case BoundKind.UnconvertedConditionalOperator:
                 case BoundKind.DefaultLiteral:
                 case BoundKind.UnconvertedInterpolatedString:
+                case BoundKind.UnconvertedCollectionLiteralExpression:
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:63]: Do we have any tests for collection literals with errors that contain some of these other expressions? In particular, we should have a test that pulls on something that will crash if it (and the nested expressions) are not successfully converted, such as IOperation.

**Review by 333fred — COMMENTED** [review]
Generally looks good to me, just one question about test coverage.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 2]
```diff
@@ -611,10 +598,34 @@ BoundExpression convertArrayElement(BoundExpression element, TypeSymbol elementT
...
+            BoundUnconvertedCollectionLiteralExpression node,
+            TypeSymbol targetType,
+            BindingDiagnosticBag diagnostics)
+        {
+            var syntax = node.Syntax;
+            var builder = ArrayBuilder<BoundExpression>.GetInstance(node.Elements.Length);
+            foreach (var element in node.Elements)
+            {
+                var result = BindToNaturalType(element, diagnostics);
+                result.WasCompilerGenerated = true;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:618]: is this needed for IOperation?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 3]
```diff
@@ -5723,7 +5732,13 @@ private BoundExpression BindUnexpectedComplexElementInitializer(InitializerExpre
             var enumeratorInfo = element.EnumeratorInfoOpt;
             if (enumeratorInfo is null)
             {
-                return element;
+                return element.Update(
+                    BindToNaturalType(element.Expression, diagnostics),
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5736]: Is this an error scenario? If so I thought there was a 'BindForErrorRecovery' method we should be using?

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 4]
```diff
@@ -1881,8 +1882,9 @@
   <Node Name="BoundCollectionLiteralExpression" Base="BoundExpression">
     <!-- Non-null type is required for this node kind -->
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
+    <Field Name="CollectionTypeKind" Type="CollectionLiteralTypeKind"/>
     <!-- An expression placeholder value representing the initialized collection. -->
-    <Field Name="Placeholder" Type="BoundObjectOrCollectionValuePlaceholder" Null="disallow" SkipInVisitor="true"/>
+    <Field Name="Placeholder" Type="BoundObjectOrCollectionValuePlaceholder?" Null="allow" SkipInVisitor="true"/>
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:1887]: generally only null in error scenarios?

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 5]
```diff
@@ -2186,17 +2186,19 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
...
                     else
                     {
-                        Debug.Assert(false); // Add test if this assert fails.
-                        (convertedType, convertedNullability) = (type, nullability);
-                        conversion = Conversion.Identity;
+                        // There was an explicit cast on top of this.
+                        (convertedType, convertedNullability) = (convertedCollection.Type, nullability);
+                        conversion = convertedCollection.CollectionTypeKind == CollectionLiteralTypeKind.None
+                            ? Conversion.NoConversion
+                            : Conversion.CollectionLiteral;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:2201]: It looks like the switch and conditional expressions above use an identity conversion in the corresponding position. What's different about this case that we use this conversion instead here?

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs [hunk 6]
```diff
@@ -348,9 +348,12 @@ static void Main()
                 """;
             var comp = CreateCompilation(source);
             comp.VerifyEmitDiagnostics(
-                // (5,13): error CS0815: Cannot assign collection literals to an implicitly-typed variable
+                // (5,17): error CS9503: There is no target type for the collection literal.
                 //         var z = [[3]];
-                Diagnostic(ErrorCode.ERR_ImplicitlyTypedVariableAssignedBadValue, "z = [[3]]").WithArguments("collection literals").WithLocation(5, 13));
+                Diagnostic(ErrorCode.ERR_CollectionLiteralNoTargetType, "[[3]]").WithLocation(5, 17),
+                // (5,18): error CS9503: There is no target type for the collection literal.
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:354]: I'm guessing there's no obvious place to hang a "target type is an error type, therefore I shouldn't error when converting for error recovery" check on the inner expression?

File: src/Compilers/Test/Core/Compilation/CompilationExtensions.cs [hunk 7]
```diff
@@ -34,11 +34,13 @@ namespace Microsoft.CodeAnalysis.Test.Utilities
     public static class CompilationExtensions
     {
         internal static bool EnableVerifyIOperation { get; } =
-#if ROSLYN_TEST_IOPERATION
-            true;
-#else
-            !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("ROSLYN_TEST_IOPERATION"));
-#endif
+            // PROTOTYPE: Re-enable
```
Comment (RikkiGibson) [inline src/Compilers/Test/Core/Compilation/CompilationExtensions.cs:37]: What's the nature of failures / work required to fix?

Comment (cston) [reply src/Compilers/Test/Core/Compilation/CompilationExtensions.cs:37]: There are 3 tests where an assert fails in `ControlFlowGraphBuilder()` in the IOperation CI leg. I'm not sure yet what is required to fix those cases, but I wanted to handle that in a specific PR that implements IOperation and CFG for collection literals.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:618]: No, we shouldn't be setting `WasCompilerGenerated` here. Removed.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:618]: Got it. I do want to make sure that disabling the IOperation tests isn't masking an actual need for this. But in either case we can address when we get around to fixing IOperation tests.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:5736]: Yes, this is an error scenario. I couldn't find a `Bind*ForErrorRecovery` method that seemed appropriate here although I might have missed one.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 4 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:1887]: Yes, this should only be `null` in error scenarios.

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs (hunk 6 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:354]: Fixed, thanks.

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:2201]: Those cases have a natural type, whereas the collection literal does not, so we need to capture whether the explicit conversion was valid (a *collection literal conversion*) or invalid (*no conversion*).

File: src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs [hunk 8]
```diff
@@ -523,14 +523,50 @@ static void Main()
...
+                        string[] y1 = [args.Length switch { 0 => null, _ => default }];
+                        string[] z1 = [args.Length == 0 ? null : default];
+                        var x2 = [(null, default)];
+                        var y2 = [args.Length switch { 0 => null, _ => default }];
+                        var z2 = [args.Length == 0 ? null : default];
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs:552]: Please pull on the semantic info in a way that will cause a crash if it survives initial binding unconverted, such as IOperation.

**Review by 333fred — APPROVED** [review]

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**